File: 206_reverse_linked_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        ListNode temp = null;
        ListNode reverseHead = head;

        while (curr != null && curr.next != null) {
            ListNode next = curr.next;

            curr.next = next.next;      // connect to third node: curr -> next -> next.next => curr.next = next.next
            next.next = reverseHead;    // connect to previous node

            reverseHead = next;         // set reverse head with new node
        }

        return reverseHead;
    }
}

File: python/206_reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList_iterative(self, head: ListNode) -> ListNode:
        # method 1. iterative
        last = None
        while head:
            temp = head.next
            head.next = last
            last = head
            head = temp
        return last

    def reverseList(self, head: ListNode) -> ListNode:
        return self.recursive(head, None)

    def recursive(self, head: ListNode, new_head: ListNode) -> ListNode:
        # method 2. recursive
        if not head: return new_head
        next = head.next
        head.next = new_head
        return self.recursive(next, head)


